fud2/fud-core/src/exec/driver.rs

```diff
@@ -16,6 +16,7 @@ pub struct Driver {
     pub ops: PrimaryMap<OpRef, Operation>,
     pub rsrc_dir: Option<Utf8PathBuf>,
     pub rsrc_files: Option<FileData>,
+    pub config_data: figment::Figment,
```

Comment:
Makes a lot of sense. The use of the configuration is to avoid having to execute the Rhai scripts twice. They have to be executed to build the driver, but they contain references to values in `config`. These have to be supplied during this first execution because I don't do a callback to Rhai type thing.

If the goal is to make the op graph invariant on config, I believe a doc comment just stating that is both more straightforward and truthful to the intent. 

Though, I am starting to think maybe two executions is better: one execution builds the op graph but doesn't actually populate the ops with their emit functions. The second execution populates these emit functions.

Reply:
The root of this problem is the ops (and thus the op graph) cannot be constructed without knowing the implementation of ops. I want to turn the Rhai code into Rust, but I don't know of a good way to just turn the Rhai code something I can call back to later. (Sam can easily implement this with closures because closures just turn into functions so the script can just call the function. Not sure how to do this with arbitrary blocks though). 

This is easily solvable in one pass by providing to the `ScriptRunner` this information. This does still have the architecture issue you mention because the stuff the ops depend on the request. 

To solve the architecture problems, either doing the callback thing currently done or multiple passes over the script is required. Alternatively, splitting the ops' implementation from the definition also would work. All of these are kind of complicated though. 

I'm of mind the simpler solution with a friendly message asking to keep the op graph invariant on requests is probably better.